Add execute-command-before BuildMutator

AlexBeggs opened the pull request:
- BuildMutator allows commands to be executed prior to the SCENARIO or BUILD phase. Enabled for all types of builds, Gradle, Bazel, Buck, and Maven.

Dependent on https://github.com/gradle/gradle-profiler/pull/315/

sunyal commented on src/main/java/org/gradle/profiler/ScenarioLoader.java:
```diff
@@ -212,80 +217,48 @@ public ScenarioLoader(GradleBuildConfigurationReader gradleBuildConfigurationRea
             int buildCount = getBuildCount(settings, scenario);
             File scenarioBaseDir = new File(settings.getOutputDir(), GradleScenarioDefinition.safeFileName(scenarioName));
 
-            if (scenario.hasPath(BAZEL) && settings.isBazel()) {
```
Refactored while I was here to make it more readable.  Extracted this into separate methods instead of a single large method

sunyal commented on src/main/java/org/gradle/profiler/ScenarioLoader.java:
```diff
@@ -212,80 +217,48 @@ public ScenarioLoader(GradleBuildConfigurationReader gradleBuildConfigurationRea
+        return scenario;
+    }
+    
+    private static Config getExecutionInstructions(File scenarioFile, InvocationSettings settings, String scenarioName, Config scenario, String toolName, List<String> toolKeys) {
```
renamed method name to be clearer

sunyal commented on src/main/java/org/gradle/profiler/ScenarioLoader.java:
```diff
@@ -298,6 +271,128 @@ private static Config getConfig(File scenarioFile, InvocationSettings settings,
         return executionInstructions;
     }
 
+    private static List<ScenarioDefinition> getScenarioDefinitions(File scenarioFile,
```
this is the section of extracted methods

sunyal commented on src/main/java/org/gradle/profiler/mutations/ExecuteCommandBuildMutator.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    public enum ExecuteCommandSchedule {
+        SCENARIO, BUILD
```
doesn't handle CLEANUP, Bazel, Buck and Maven don't implement that

sunyal commented on src/main/java/org/gradle/profiler/mutations/ProcessBuilderCommandInvoker.java:
```diff
@@ -0,0 +1,24 @@
+                .redirectOutput(ProcessBuilder.Redirect.INHERIT)
+                .redirectError(ProcessBuilder.Redirect.INHERIT);
+            Process process = processBuilder.start();
+            return process.waitFor();
```
could potentially add a timeout but for now I left it

sunyal commented on src/main/java/org/gradle/profiler/ScenarioLoader.java:
```diff
@@ -212,80 +218,26 @@ public ScenarioLoader(GradleBuildConfigurationReader gradleBuildConfigurationRea
             String title = scenario.hasPath(TITLE) ? scenario.getString(TITLE) : null;
 
             List<BuildMutator> mutators = BUILD_MUTATOR_CONFIGURATORS.entrySet().stream()
-                .filter(entry -> scenario.hasPath(entry.getKey()))
```
filtering isn't done here. The BuildMutator's should be responsible for determining if they support the configuration or not.

AlexBeggs committed:
Add execute-command-before BuildMutator
- BuildMutator allows commands to be executed prior to the SCENARIO or BUILD phase. Enabled for all types of builds, Gradle, Bazel, Buck, and Maven.

Signed-off-by: Alex Beggs <sunyal@gmail.com>

ZacSweers commented:
Does this run commands in the root directory of the project? It would be helpful I think to have an example that runs a script that is local to the project being profiled since I think that would be a common case.

AlexBeggs commented:
If I recall correctly, this runs the command in the working directory or `user.dir`
